### PR TITLE
Add Triton NVFP4 attention QAT backward

### PR DESCRIPTION
**Pull request by SolitaryThinker**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] Maintainer action: add `cat-feature`, `mod-cutedsl`, `mod-frontend`, `orig-external`, and `op:sdpa` (external contributors cannot assign labels).
- [x] I set the **Milestone** and **Projects** fields in the sidebar (required to merge; maintainers can set these for external contributions).

## Affected area

FE OSS kernels or CuTeDSL

## Summary

- Adds an experimental Triton NVFP4 QAT SDPA backward API with the explicit `APIBase` lifecycle and a convenience wrapper.
- Ports FastVideo's local-scale Q/K/V/P fake quantization and split dQ/dK-dV topology, including the SM100 64x64 production path and Blackwell fallback paths.
- Adds lazy exports, optional Triton packaging, numerical and support tests, a benchmark, documentation, and third-party notices.

The implementation is based on the FastVideo work in [PR #1460](https://github.com/hao-ai-lab/FastVideo/pull/1460) and [PR #1619](https://github.com/hao-ai-lab/FastVideo/pull/1619), pinned in `THIRD_PARTY_LICENSES.txt` to the source revision used for this port.

## Why

FastVideo's training path needs explicit gradients that match NVFP4 fake quantization and straight-through-estimator semantics. Keeping fake quantization, the saved-forward auxiliary contract, workspace ownership, and the split backward kernels together provides a reusable FE OSS surface while preserving allocation-free, stream-aware execution.

## Related issues

None.

## API and compatibility impact

This adds the experimental `Nvfp4AttentionQatBackward` class and `nvfp4_attention_qat_backward` wrapper. The initial contract supports contiguous 16-byte-aligned BF16 MHA tensors with head dimension 128, natural-log LSE, SM100/SM103/SM120/SM121, noncausal self or cross attention, and causal equal-length attention. It requires the optional `triton` extra (Triton 3.7 or newer on Linux with Python 3.10 or newer) and a CUDA-enabled PyTorch installation. Existing APIs are unchanged.

## Testing

- `uvx pre-commit run` over all changed files: passed.
- `python3 -m compileall -q python/cudnn/sdpa/bwd/qat benchmark/attention_training/benchmark_nvfp4_qat_backward.py test/python/fe_api/sdpa/test_nvfp4_attention_qat_backward.py`: passed.
- `git diff --check origin/develop...HEAD`: passed.
- Triton warmup compile matrix with PyTorch 2.13.0 and Triton 3.7.1: all five specializations compiled for SM100, SM103, SM120, and SM121, covering causal and noncausal fallback configurations and the SM100 64x64 production configuration.
- `cd test/python && PYTHONPATH=. uv run --no-project --with torch==2.13.0 --with triton==3.7.1 --with pytest --with numpy pytest --noconftest fe_api/sdpa/test_nvfp4_attention_qat_backward.py -q -m 'L0 or L1'`: collected successfully; six tests skipped because no CUDA device was available.

GPU numerical and performance execution was not run because this environment has no NVIDIA GPU or built local cuDNN extension. The new tests and benchmark are included for Blackwell CI and hardware validation.

## Summary by CodeRabbit

* **New Features**
  * Added experimental NVFP4 quantization-aware attention backward support for compatible Blackwell GPUs.
  * Added APIs for computing attention gradients with optional caller-provided output buffers.
  * Added validation for tensor layouts, data types, causal shapes, workspace, and hardware compatibility.
  * Added optional Triton installation support.

* **Documentation**
  * Documented API usage, requirements, limitations, licensing, and benchmark instructions.

* **Tests**
  * Added coverage for accuracy, causal and variable-length inputs, workspace handling, and invalid configurations.

* **Benchmarks**
  * Added a CUDA benchmark for NVFP4 QAT backward performance.

### CODE REVIEW

File: docs/fe-oss-apis/attention/nvfp4_attention_qat_backward.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+non-causal SM100 configuration uses 64 by 64 tiles; other supported Blackwell
+configurations use 32 by 32 tiles.
+
+## Installation
+
+From a source checkout, install the CuTe DSL base dependencies, Triton, and the
+torch dependency group:
+
+```bash
+pip install -e ".[cutedsl,triton]" --group torch
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply docs/fe-oss-apis/attention/nvfp4_attention_qat_backward.md]: Addressed in 69e211ac. The source-install instructions now upgrade to pip 25.1 or newer before the unchanged editable install command that uses --group.

File: python/cudnn/sdpa/bwd/qat/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+"""NVFP4 quantization-aware SDPA backward."""
+
+from .api import Nvfp4AttentionQatBackward, nvfp4_attention_qat_backward
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply python/cudnn/sdpa/bwd/qat/__init__.py]: Addressed in 69e211ac. The QAT package now resolves both public symbols through module-level __getattr__, preserves __all__, raises AttributeError for unknown names, and has an import-boundary regression test proving the namespace does not load torch or Triton.

File: python/cudnn/sdpa/bwd/qat/api.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        if current_stream is None:
+            yield
+            return
+        stream_handle = int(current_stream)
+        torch_current = torch.cuda.current_stream(device)
+        if stream_handle in (0, 1, 2) or stream_handle == torch_current.cuda_stream:
+            yield
+            return
+        torch_default = torch.cuda.default_stream(device)
+        launch_stream = torch_default if stream_handle == torch_default.cuda_stream else torch.cuda.ExternalStream(stream_handle, device=device)
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply python/cudnn/sdpa/bwd/qat/api.py:35]: Addressed in 69e211ac. Only a handle equal to the active PyTorch stream now takes the no-op path; explicit default-stream handles are routed through torch.cuda.stream so the Triton launches use the requested stream.

**Comment by SolitaryThinker** [comment]
Review follow-up in 69e211ac:

- Bounded the shape-specialized operator cache to 64 entries with LRU refresh and eviction.
- Moved workspace offsets, dtypes, shapes, and total-size calculation into one shared layout helper used by both support planning and runtime carving.
- Added docstrings across the touched functions; whole-file coverage for the changed Python files is now 40/46 (87%), with every function in the NVFP4 QAT files documented.
- Added a regression check that importing the QAT package namespace does not load torch or Triton.

Validation completed for the update:

- staged pre-commit hooks: passed
- Python compileall and diff checks: passed
- focused workspace-layout, lazy-import, stream-routing, and bounded-LRU checks: passed
- Triton warmup compilation: all five artifacts compiled for SM100, SM103, SM120, and SM121 across production, fallback, causal, and noncausal configurations
- focused pytest collection: six GPU tests collected and skipped because this host has no CUDA device

File: docs/fe-oss-apis/attention/nvfp4_attention_qat_backward.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+non-causal SM100 configuration uses 64 by 64 tiles; other supported Blackwell
+configurations use 32 by 32 tiles.
+
+## Installation
+
+From a source checkout, install the CuTe DSL base dependencies, Triton, and the
+torch dependency group:
+
+```bash
+python -m pip install --upgrade "pip>=25.1"
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply docs/fe-oss-apis/attention/nvfp4_attention_qat_backward.md:42]: Addressed in 3d0d2586. Both source-install commands now use python -m pip, so the upgrade and editable installation target the same interpreter.

**Comment by SolitaryThinker** [comment]
The callable-lifecycle observation is already covered by the inherited APIBase.__call__ implementation in python/cudnn/api_base.py: it invokes compile when _compiled_kernel is absent and then delegates all arguments to execute. Commit 3d0d2586 updates the class API test to call the operator directly without an explicit compile, confirming that inherited first-call compilation path while avoiding a duplicate subclass implementation.

File: test/python/fe_api/sdpa/test_nvfp4_attention_qat_backward.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    q, k, v, high_precision_o, do, lse, scale = inputs
+    op = Nvfp4AttentionQatBackward(q, k, v, high_precision_o, do, lse, softmax_scale=scale)
+    assert op.check_support()
+
+    dq = torch.empty_like(q)
+    dk = torch.empty_like(k)
+    dv = torch.empty_like(v)
+    workspace = torch.empty(op.scratch_workspace_bytes(), dtype=torch.uint8, device=q.device)
+    op(q, k, v, high_precision_o, do, lse, dq, dk, dv, workspace)
+    assert op._compiled_kernel is not None
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply test/python/fe_api/sdpa/test_nvfp4_attention_qat_backward.py]: Addressed in 0d48d41c. The direct-call test now retains all three reference gradients, initializes dQ/dK/dV to NaN, and compares every produced buffer against the reference with the tail/cross-attention tolerance while retaining the support and compiled-kernel assertions.

**Review by YangXu1990uiuc — APPROVED** [review]

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run python_tests
